### ISSUE TITLE
Executing a plugin with mocked dependent class

### ISSUE THREAD

**Issue by LeeJP1983** [jordimontana82/fake-xrm-easy#498]
I have a plugin which calls a class that runs some implementation. 

ClassA : ClassB
{
  DoSomething()
}

MyPlugin : IPlugin
{
    Execute()
    {
        ClassA class = new ClassA()
        class.DoSomething();
    }
}

I would like to mock ClassA so that DoSomething returns something different for testing. How do I tell FakeXrmEasy to use this mocked Class during the plugin context instead of the real one?

Thanks

**Comment by jordimontana82** [comment]
Hi @LeeJP1983 , as the plugin constructor doesn't allow to inject any other dependencies, I usually put these dependencies as properties (i.e. usually a property exposing an interface), then using the ExecutePluginWith overload that allows to pass you a specific plugin instance with the mocked class / interface...

**Comment by LeeJP1983** [comment]
Thanks @jordimontana82. 

I was sort of able to get it to work, but perhaps it's not working 100% because of my inexperience with interfaces.

If I remove ClassA class = new ClassA(), I have no problem with the unittest. But obviously I need to construct ClassA in the Execute function.

**Comment by jordimontana82** [comment]
Exactly, I pass in the mock interface as a parameter, but the real execution will use a new instance of ClassA inside the execute method. Something along the lines of:

    MyPlugin : IPlugin
    {
      public ClassA myDependency {get; set;}
      Execute()
      {
          /// ...
          if(myDepedency == null) 
          {
                myDependency = new ClassA()
          }
     
          myDependency.DoSomething();
      }
   }


So the plugin will use the "injected" class from the unit test project or else the real one if no one was defined. I'm closing this one now, any other questions let me know.

**Comment by LeeJP1983** [comment]
Hey @jordimontana82,

Let me make this slightly more complex.

`
abstract ClassB()
{
    Property1 {get; set;}
    DoSomething(){}
}
ClassA : ClassB
{
    ClassA(var1, var2)
    {
        orgService.Retrieve()
        Property1 = DoSomething()
    }
}
MyPlugin : IPlugin
{
    public ClassA MyDependency {get; set;}
    Execute()
    {
        if(myDepedency == null) 
        {
              myDependency = new ClassA()
        }
    }
}
`

I feel the issue I'm running into has to do with me calling DoSomething() inside the constructor for ClassA, but I need to call it to set a property. I'd prefer to do this call inside ClassA. What I am running into when I do this is that when I call this in the unittest

`var myPlugin = new MyPlugin(unsecure, secure) { MyDependency = new Mock<ClassA>([var1, var2])`

It will run through the ClassA constructor before I've called `ExecutePluginWith(pluginContext, myPlugin)`

Or do I need to refactor this to pull those dependencies out of the ClassA constructor?

**Comment by jordimontana82** [comment]
I'd say better to move the DoSomething() call outside of the constructor yes, or else , at least, the constructor should have an extra IOrganizationService parameter , which you could pass in your mock or the real OrganizationService from the plugin execution context when doing:

        if(myDepedency == null) 
        {
              myDependency = new ClassA(service);  //service form plugin execution context here
        }

**Comment by jordimontana82** [comment]
Basically, you'll need to surface that IOrganizationService depedency all the way up for you to inject a mock or a real service...

**Comment by LeeJP1983** [comment]
Thanks Jordi,

Currently I have the IOrganizationService being passed in to the ClassA constructor and stored as a property in the abstract ClassB.

My DoSomething() constructs a RetrieveAttributeRequest and passes that into OrganizationService.Execute(). I've also setup the EntityMetadata correctly in the unittest, as this all worked before I tried to abstract this into a class. I've called InitializeMetadata before constructing ClassA but the Faked OrganizationService doesn't seem to have visibility on this metadata until I execute the plugin.

Any thoughts?
